code/krepel/math/vector3.d

```diff
@@ -0,0 +1,569 @@
+{
+  return Vec - Vec.ProjectOntoNormal(Normal);
+}
+
+/// Reflects a Vector around a Normal (Normal needs to be Normalized)
+/// Returns the reflected Vector
+/// Input Vectors will not be modified
+Vector3 ReflectVector(Vector3 Vec, Vector3 Normal)
+{
+  return Vec - (2 * (Vec | Normal) * Normal);
+}
+
+/// Checks if any component inside the Vector is NaN.
+/// Input Vector will not be modified
+bool ContainsNaN(Vector3 Vec)
+{
+  return IsNaN(Vec.X) || IsNaN(Vec.Y) || IsNaN(Vec.Z);
+}
+
+/// Checks if two Vectors are nearly equal (are equal with respect to a scaled epsilon)
+/// Input Vectors will not be modified
+bool NearlyEquals(Vector3 A, Vector3 B, float Epsilon = 1e-4f)
+{
+  return krepel.math.NearlyEquals(A.X, B.X, Epsilon) &&
+         krepel.math.NearlyEquals(A.Y, B.Y, Epsilon) &&
+         krepel.math.NearlyEquals(A.Z, B.Z, Epsilon);
+}
+
+/// Returns a clamped copy of the given Vector
+/// The retuned Vector will be of size <= MaxSize
+/// Input Vector will not be modified
+Vector3 ClampSize(Vector3 Vec, float MaxSize)
+{
+  Vector3 Normal = Vec.NormalizedCopy();
+  return Normal * Min(Vec.Length(), MaxSize);
+}
+
+/// Returns a clamped copy of the X and Y component of the Vector, the Z compnent will be untouched
+/// The length of the X and Y component will be <= MaxSize
+/// Input Vector will not be modified
+Vector3 ClampSize2D(Vector3 Vec, float MaxSize)
+{
+  Vector3 Clamped = Vec;
+  Clamped.Z = 0;
+  Clamped.Normalize();
+  Clamped *= Min(Vec.Length2D(), MaxSize);
+  Clamped.Z = Vec.Z;
+  return Clamped;
+}
+
+struct Vector3
+{
+  @safe:
+  @nogc:
+  nothrow:
+  union
+  {
+    struct
+    {
+      float X, Y, Z;
+    }
+    float[3] Data;
+  }
+  this(float Value)
+  {
+    X = Value;
+    Y = Value;
+    Z = Value;
+  }
+
+  this(float X, float Y, float Z)
+  {
+    this.X = X;
+    this.Y = Y;
+    this.Z = Z;
+  }
+
+  this(Vector2 Vec, float Z)
+  {
+    this.Data[0..2] = Vec.Data[];
+    this.Z = Z;
+  }
+
+  this(float X, Vector2 Vec)
+  {
+    this.X = X;
+    this.Data[1..3] = Vec.Data[];
+  }
+
+  /// Normalizes the Vector (Vector will have a length of 1.0)
+  void Normalize()
+  {
+    // Don't return result to avoid confusion with NormalizedCopy
+    // and stress that this operation modifies the Vector on which it is called
+    float Length = this.Length();
+    this /= Length;
+  }
+
+  // Dot product
+  float opBinary(string Op:"|")(Vector3 Rhs)
+  {
+    return
+      X * Rhs.X +
+      Y * Rhs.Y +
+      Z * Rhs.Z;
+  }
+
+  Vector3 opOpAssign(string Operator)(float Rhs)
+  {
+    static if(Operator == "*" || Operator == "/")
+    {
+      auto Result = mixin("Vector3("~
+        "X" ~ Operator ~ "Rhs,"
+        "Y" ~ Operator ~ "Rhs,"
+        "Z" ~ Operator ~ "Rhs)");
+      Data = Result.Data;
+      return this;
+    }
+    else
+    {
+      static assert(false, "Operator " ~ Operator ~ " not implemented.");
+    }
+  }
+
+  //const (char)[] ToString() const
```

Comment:
I think we will do these kinds of things in a much more general way using compile-time reflection at some other point.